### PR TITLE
[Core][Kernels] Enable FP8 KV Cache with Flashinfer backend.  + BugFix for kv_cache_dtype=auto

### PR DESCRIPTION
**Pull request by pavanimajety**
Previous PR with dicussions: https://github.com/vllm-project/vllm/pull/7798

Revert PR due to bug with `--kv-cache-dtype=auto` : https://github.com/vllm-project/vllm/pull/7982

`--kv-cache-dtype=fp8`
```
model_str="neuralmagic/Meta-Llama-3-8B-Instruct-FP8"
model = LLM(model=model_str, quantization="fp8",kv_cache_dtype="fp8")
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "New york times is politically sided to ",
]
```
Results:
```
$ VLLM_ATTENTION_BACKEND=FLASHINFER /bin/python3 /workspace/vllm_github/test_llm.py
..........

 Prompt: 'Hello, my name is', 
Generated text: ' [Your Name]. I am a [Your Profession/Student] and I am', 
token_ids: array('l', [510, 7927, 4076, 948, 358, 1097, 264, 510, 7927, 50311, 14, 14428, 60, 323, 358, 1097])


 Prompt: 'The president of the United States is', 
Generated text: ' the head of state and the head of government of the United States. The president', 
token_ids: array('l', [279, 2010, 315, 1614, 323, 279, 2010, 315, 3109, 315, 279, 3723, 4273, 13, 578, 4872])


 Prompt: 'The capital of France is', 
Generated text: ' Paris, which is located in the north-central part of the country. Paris is', 
token_ids: array('l', [12366, 11, 902, 374, 7559, 304, 279, 10411, 85181, 961, 315, 279, 3224, 13, 12366, 374])


 Prompt: 'New york times is politically sided to ', 
Generated text: '1 side\nThe New York Times is often accused of being biased towards the left', 
token_ids: array('l', [16, 3185, 198, 791, 1561, 4356, 8691, 374, 3629, 13487, 315, 1694, 48761, 7119, 279, 2163])

```

`--kv-cache-dtype=auto`

```
model_str="neuralmagic/Meta-Llama-3-8B-Instruct-FP8"
model = LLM(model=model_str, quantization="fp8",kv_cache_dtype="auto")
```

Results: 

```
/bin/python3 /workspace/vllm_github/test_llm.py

........

Processed prompts: 100%|███████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 36.28it/s, est. speed input: 254.03 toks/s, output: 580.62 toks/s]


 Prompt: 'Hello, my name is', 
Generated text: ' [Name]. I am a [Your Profession/Student] and I am here', 
token_ids: array('l', [510, 678, 948, 358, 1097, 264, 510, 7927, 50311, 14, 14428, 60, 323, 358, 1097, 1618])


 Prompt: 'The president of the United States is', 
Generated text: ' the head of state and head of government of the United States. The president is', 
token_ids: array('l', [279, 2010, 315, 1614, 323, 2010, 315, 3109, 315, 279, 3723, 4273, 13, 578, 4872, 374])


 Prompt: 'The capital of France is', 
Generated text: ' Paris, which is located in the north-central part of the country. Paris is', 
token_ids: array('l', [12366, 11, 902, 374, 7559, 304, 279, 10411, 85181, 961, 315, 279, 3224, 13, 12366, 374])


 Prompt: 'New york times is politically sided to ', 
Generated text: '1 side\nThe New York Times is often accused of being biased towards the left', 
token_ids: array('l', [16, 3185, 198, 791, 1561, 4356, 8691, 374, 3629, 13487, 315, 1694, 48761, 7119, 279, 2163])

```

### CODE REVIEW

**Comment by pavanimajety** [comment]
@simon-mo  @youkaichao  @comaniac  Can one of you please take a look? 

Thanks!

File: vllm/attention/backends/flashinfer.py [hunk 1]
```diff
@@ -83,6 +83,15 @@ def copy_blocks(
...
         return [64, 128, 256]
 
+    @staticmethod
+    def get_fp8_dtype_for_flashinfer(kv_cache_dtype: str) -> torch.dtype:
+        if kv_cache_dtype in ("fp8", "fp8_e4m3"):
+            return torch.float8_e4m3fn
+        elif kv_cache_dtype == "fp8_e5m2":
+            return torch.float8_e5m2
+        else:
+            return ValueError(f"Unrecognized FP8 dtype: {kv_cache_dtype}")
```
Comment (youkaichao) [inline vllm/attention/backends/flashinfer.py:93]: should be `raise`

File: vllm/attention/backends/flashinfer.py [hunk 2]
```diff
@@ -674,6 +684,12 @@ def forward(
                 k_scale,
                 v_scale,
             )
+            # The FlashInfer api requires data to be in fp8_e4m3 or fp8_e5m2
+            # to process the cache when the kv_cache_dtype is fp8
+            if self.kv_cache_dtype in ["fp8", "fp8_e4m3", "fp8_e5m2"]:
```
Comment (comaniac) [inline vllm/attention/backends/flashinfer.py:689]:
```suggestion
            if self.kv_cache_dtype.startswith("fp8"):
```

File: tests/kernels/test_flashinfer.py [hunk 3]
```diff
@@ -249,3 +253,215 @@ def test_flashinfer_prefill_with_paged_kv(seq_lens: List[Tuple[int, int]],
...
         f"{torch.max(torch.abs(output - ref_output))}"
+
+
+@pytest.mark.parametrize("seq_lens", [[(1, 132), (5, 18)]])
+@pytest.mark.parametrize("num_heads", [(32, 8), (6, 1)])
+@pytest.mark.parametrize("head_size", HEAD_SIZES)
+@pytest.mark.parametrize("block_size", BLOCK_SIZES)
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("soft_cap", [None, 30.0, 50.0])
+def test_flashinfer_prefill_with_paged_fp8_kv(
```
Comment (comaniac) [inline tests/kernels/test_flashinfer.py:264]: To avoid the bug in the future, do we have a unit test for FlashInfer with kv cache dtype "auto"?

**Comment by pavanimajety** [comment]
Can you please help with the CI here? It seems to be yellow without logs

File: vllm/attention/backends/flashinfer.py [hunk 4]
```diff
@@ -340,7 +349,7 @@ def begin_forward(self):
                 self.page_size,
                 # Disable flashinfer's pos encoding and use vllm's rope.
                 pos_encoding_mode="NONE",
-                data_type=self.data_type)
+            )
```
Comment (mgoin) [inline vllm/attention/backends/flashinfer.py:352]: Why is this not needed anymore?

**Review by mgoin — APPROVED** [review]
LGTM thanks for the kernel tests! Just one question about the data_type arg to flashinfer

File: vllm/attention/backends/flashinfer.py (hunk 4 above)
Comment (pavanimajety) [reply vllm/attention/backends/flashinfer.py:352]: The wrappers are throwing dispatch errors with Float8 datatype with the current flashinfer versions(0.1.4 and 0.1.5). I have a couple of future MRs planned. I will enable it if future releases of Flashinfer fix the dispatch api. Seems like there would be api name changes with flashinfer too.

Comment (mgoin) [reply vllm/attention/backends/flashinfer.py:352]: I see, this is sufficient for now and it seems we will likely need to address this in flashinfer first

**Comment by pavanimajety** [comment]
Follow-up: https://github.com/vllm-project/vllm/issues/8009

**Pull request merged**